### PR TITLE
Combine CoreSpan with AgentSpan

### PR DESCRIPTION
**Pull request by tylerbenson**
Because CoreSpan was in core, it could not be used by instrumentation.  With AgentSpan being the primary API, it can be used in instrumentation.

I also cleaned up some other duplicate/unnecessary methods in the process. (`spanName` getter/setter)

### CODE REVIEW

**Review by richardstartin — CHANGES_REQUESTED** [review]
I don't understand the rationale for this

File: dd-trace-core/src/main/java/datadog/trace/common/metrics/MetricsAggregator.java [hunk 1]
```diff
@@ -1,10 +1,10 @@
...
 
-import datadog.trace.core.CoreSpan;
+import datadog.trace.bootstrap.instrumentation.api.AgentSpanData;
 import java.util.List;
 
 public interface MetricsAggregator extends AutoCloseable {
   void start();
 
-  void publish(List<? extends CoreSpan<?>> trace);
+  void publish(List<? extends AgentSpanData> trace);
```
Comment (richardstartin) [inline dd-trace-core/src/main/java/datadog/trace/common/metrics/MetricsAggregator.java:9]: This is a constraining change

File: dd-trace-core/src/main/java/datadog/trace/common/metrics/ConflatingMetricsAggregator.java [hunk 2]
```diff
@@ -86,17 +86,17 @@ public void start() {
   }
 
   @Override
-  public void publish(List<? extends CoreSpan<?>> trace) {
+  public void publish(List<? extends AgentSpanData> trace) {
     if (enabled) {
-      for (CoreSpan<?> span : trace) {
+      for (AgentSpanData span : trace) {
```
Comment (richardstartin) [inline dd-trace-core/src/main/java/datadog/trace/common/metrics/ConflatingMetricsAggregator.java:91]: Reduces flexibility

**Comment by richardstartin** [comment]
The reason this interface exists at the moment is to break dependencies between sampling/metrics/processing and `DDSpan`/`DDSpanContext`, which I think are in need of refactoring. The reason for the recursive type bound is that I think this will probably be useful in the future, but hard to add later. The interface isn't the final solution, but it's not useless. It's also not a stable interface, and I don't really want it escaping into our API, or for our own instrumentation to create impediments to changing it at the moment, so I'm not in favour of the changes here.

**Comment by tylerbenson** [comment]
I'm going to close this for now and revisit when it's more important.

**Pull request closed without merging**